lxd/storage/drivers/drivers_lvm_utils.go

```diff
@@ -0,0 +1,693 @@
+
+			revert.Add(func() {
+				d.removeLogicalVolume(d.lvmDevPath(d.config["lvm.vg_name"], newSnapVol.volType, newSnapVol.contentType, newSnapVol.name))
+			})
+		}
+	}
+
+	// Handle copying the main volume.
+	if d.HasVolume(vol) {
+		if refresh {
+			newVolDevPath := d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, vol.name)
+			tmpVolName := fmt.Sprintf("%s%s", vol.name, tmpVolSuffix)
+			tmpVolDevPath := d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, tmpVolName)
+
+			// Rename existing volume to temporary new name so we can revert if needed.
+			err := d.renameLogicalVolume(newVolDevPath, tmpVolDevPath)
+			if err != nil {
+				return fmt.Errorf("Error temporarily renaming original LVM logical volume: %v", err)
+			}
+
+			// Record this volume to be removed at the very end.
+			removeVols = append(removeVols, tmpVolName)
+
+			revert.Add(func() {
+				// Rename the original volume back to the original name.
+				d.renameLogicalVolume(tmpVolDevPath, newVolDevPath)
+			})
+		} else {
+			return fmt.Errorf("LVM volume already exists: %s", vol.name)
+		}
+	} else {
+		volPath := vol.MountPath()
+		err := vol.EnsureMountPath()
+		if err != nil {
+			return err
+		}
+
+		revert.Add(func() { os.RemoveAll(volPath) })
+	}
+
+	// Create snapshot of source volume as new volume.
+	_, err := d.createLogicalVolumeSnapshot(d.config["lvm.vg_name"], srcVol, vol, false, d.usesThinpool())
+	if err != nil {
+		return fmt.Errorf("Error creating LVM logical volume snapshot: %v", err)
+	}
+
+	volDevPath := d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, vol.name)
+
+	revert.Add(func() {
+		d.removeLogicalVolume(volDevPath)
+	})
+
+	if vol.contentType == ContentTypeFS {
+		// Generate a new filesystem UUID if needed (this is required because some filesystems won't allow
+		// volumes with the same UUID to be mounted at the same time). This should be done before volume
+		// resize as some filesystems will need to mount the filesystem to resize.
+		if renegerateFilesystemUUIDNeeded(d.volumeFilesystem(vol)) {
+			d.logger.Debug("Regenerating filesystem UUID", log.Ctx{"dev": volDevPath, "fs": d.volumeFilesystem(vol)})
+			err = regenerateFilesystemUUID(d.volumeFilesystem(vol), volDevPath)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Resize the new volume and filesystem to the correct size.
+	err = d.SetVolumeQuota(vol, d.volumeSize(vol), nil)
+	if err != nil {
+		return err
+	}
+
+	// Finally clean up original volumes left that were renamed with a tmpVolSuffix suffix.
+	for _, removeVolName := range removeVols {
+		err := d.removeLogicalVolume(d.lvmDevPath(d.config["lvm.vg_name"], vol.volType, vol.contentType, removeVolName))
+		if err != nil {
+			return errors.Wrapf(err, "Error removing LVM volume: %s", vol.name)
+		}
+	}
+
+	revert.Success()
+	return nil
+}
+
+// logicalVolumeSize gets the size in bytes of a logical volume.
+func (d *lvm) logicalVolumeSize(volDevPath string) (int64, error) {
```

Comment:
Only has a single caller, any reason to keep it separate?

Reply:
@stgraber yep, the same reason as `volumeGroupExtentSize`, my intention is to avoid distracting the reader of `SetVolumeQuota` with boiler plate parsing code, and allow them to focus on the resize logic.